crates/ide-assists/src/handlers/recursively_add_derive.rs

```diff
@@ -0,0 +1,1010 @@
+        let macro_ = ctx.sema.resolve_path(&path)?.as_module_def()?.as_macro()?;
+        SmallVec::from([DeriveItem::from_path(ctx, current_crate, &adt_ty, macro_, name)?])
+    };
+
+    let to_derive = field_types_to_derive(ctx, &adt_src, &derive_items, &current_crate)?;
+    if to_derive.iter().all(|(_, derives)| derives.is_empty()) {
+        return None;
+    }
+
+    let formatted_items = derive_items.iter().try_fold(String::new(), |mut acc, item| {
+        let name = item.derive_macro.name(ctx.db());
+        let name = name.display(ctx.db(), current_edition);
+        if acc.is_empty() {
+            write!(acc, "{name}").ok()?;
+        } else {
+            write!(acc, ", {name}").ok()?;
+        }
+        Some(acc)
+    })?;
+    acc.add(
+        AssistId("recursively_add_derive", AssistKind::Generate),
+        format!(
+            "Recursively add `#[derive({formatted_items})]` on each field type within this crate"
+        ),
+        adt_src.syntax().text_range(),
+        |edit| {
+            for (src, derives) in to_derive {
+                let mut adt_src = src.value;
+                let Some(module) = ctx.sema.scope(adt_src.syntax()).map(|scope| scope.module())
+                else {
+                    continue;
+                };
+
+                // Allow for multi-file edits.
+                let Some(file_id) = src.file_id.file_id() else { continue };
+                edit.edit_file(file_id);
+
+                // Create a comma-separate list containing paths to derive items.
+                let derives_len = derives.len();
+                let derive_paths = derives
+                    .into_iter()
+                    .filter_map(|item| {
+                        module.find_path(ctx.db(), item, ctx.config.import_path_config())
+                    })
+                    .map(|path| mod_path_to_ast(&path, current_edition))
+                    .enumerate()
+                    .flat_map(|(i, path)| {
+                        let add_comma_space = i != 0 && i < derives_len;
+                        add_comma_space
+                            .then(|| {
+                                [make::token(T![,]).into(), make::tokens::single_space().into()]
+                            })
+                            .into_iter()
+                            .flatten()
+                            .chain(path.syntax().clone_for_update().descendants_with_tokens())
+                    });
+
+                let maybe_derive_attr = adt_src.attrs().find_map(|attr| {
+                    if attr.path()?.syntax().text() == "derive" {
+                        attr.token_tree()
+                    } else {
+                        None
+                    }
+                });
+
+                if let Some(derive_attr) = maybe_derive_attr {
+                    // Append to the first existing `#[derive]` attribute.
+                    let derive_attr = edit.make_mut(derive_attr);
```

Comment:
After a bit of experimenting I have a few questions:

* Using `SyntaxEditor::new(adt_src.syntax().clone())` [here](https://github.com/rust-lang/rust-analyzer/pull/18118/commits/f6b20fd238b2da46def9192b686321d1af27f663#diff-73d26aef55f4fe15867a3e22d32e18c45c6750cbdbba975b70b198fc6710bed3R111) appears to cause a crash when calling `SyntaxEditor::finish`: 
  ```
  can't resolve SyntaxNodePtr { kind: STRUCT, range: 0..11 } with STRUCT@0..11
  ```
  It works fine when using the parent of `adt_src.syntax()` though, which seemingly implies that you are required to pass in the files root syntax node. Does this mean you need a separate `SyntaxEditor` for each file you want to edit? 
* How should the resulting `SyntaxEdit` from `SyntaxEditor::finish` be applied inside of an assist?

Reply:
- Yeah, you generally want to make a separate `SyntaxEditor` for each file you want to edit. Creating `SyntaxEditor`s is generally lightweight, as they're mostly just the collection of edits you want to perform. It's also definitely recommended to use the root syntax node so that other edits in a file can be composed easily.
- Currently, it's a little tricky, as you'll need to `make_mut` the root syntax node and replace it with `SyntaxEdit::new_root`.

#18134 makes it easier to work with `SyntaxEditor`s and also includes an example of a before/after with `SyntaxEditor`. Unfortunately you won't get all of the niceties of the current mutable syntax trees yet (e.g. `add_attr`), so for something like that you'll need to do
```rust
let new_adt = adt_src.clone_for_update().add_attr(derive);
edit.replace(adt_src.syntax(), new_adt.syntax());
```